### PR TITLE
Add check for valid URL/PORT in env.yaml if PORT defined in manifest

### PR DESCRIPTION
**Pull request by groeges**
Signed-off-by: Steven Groeger <groeges@uk.ibm.com>

**What type of PR is this?**
/kind bug

> Documentation changes: Please include [skip ci] in your commit message as well
> /kind documentation
> [skip ci]

**What does does this PR do / why we need it**:
If there is a PORT substitution variable in the deployment-manifest being used, then check there is a valid URL with PORT defined in env.yaml (as a result of running `odo url create`). If not, then fail with suitable error.

**Which issue(s) this PR fixes**:

Fixes #?

**How to test changes / Special notes to the reviewer**:

### CODE REVIEW

File: pkg/odo/cli/component/deploy.go [hunk 1]
```diff
@@ -79,6 +76,13 @@ func (do *DeployOptions) Complete(name string, cmd *cobra.Command, args []string
 func (do *DeployOptions) Validate() (err error) {
 
 	log.Infof("\nValidation")
+
+	envInfo, err := envinfo.NewEnvSpecificInfo(do.componentContext)
+	if err != nil {
+		return errors.Wrap(err, "unable to retrieve configuration information")
+	}
+	do.EnvSpecificInfo = envInfo
+
```
Comment (EnriqueL8) [inline pkg/odo/cli/component/deploy.go]: Any reason why we moved this to the `Validate()` function? I followed the same patterns as the other commands to put it in `Complete()`

File: pkg/odo/cli/component/deploy.go [hunk 2]
```diff
@@ -148,6 +152,23 @@ func (do *DeployOptions) Validate() (err error) {
...
+	// check if manifestSource contains PORT template variable
+	// if it does, then check we have an port setup in env.yaml
+	if bytes.Contains(manifestBytes, []byte("PORT")) {
+		deploymentPort := 0
+		for _, localURL := range do.EnvSpecificInfo.GetURL() {
+			if localURL.Kind == envinfo.ROUTE {
+				deploymentPort = localURL.Port
+				break
+			}
+		}
```
Comment (EnriqueL8) [inline pkg/odo/cli/component/deploy.go]: should we be using the determine port function here? We could move the function from the adapter to utils and then set the result into `DeployOptions` struct

Comment (groeges) [reply pkg/odo/cli/component/deploy.go]: We could do that. Will investigate and update PR if it works.

File: pkg/odo/cli/component/deploy.go (hunk 1 above)
Comment (groeges) [reply pkg/odo/cli/component/deploy.go]: My misunderstanding of when Complete() and Validate() get called. Backing out this part of change.

File: pkg/odo/cli/component/deploy.go (hunk 2 above)
Comment (groeges) [reply pkg/odo/cli/component/deploy.go]: Have moved to pkg/envinfo/envinfo.go

File: pkg/envinfo/envinfo.go [hunk 3]
```diff
@@ -297,6 +298,16 @@ func (ei *EnvInfo) GetDebugPort() int {
 	return *ei.componentSettings.DebugPort
 }
 
+// GetPortByURLKind returns the Port of a specific URL type, returns 0 if nil
+func (ei *EnvInfo) GetPortByURLKind(urlKind URLKind) (int, error) {
```
Comment (EnriqueL8) [inline pkg/envinfo/envinfo.go:302]: nice!

**Review by EnriqueL8 — APPROVED** [review]
Looks good!

**Pull request merged**